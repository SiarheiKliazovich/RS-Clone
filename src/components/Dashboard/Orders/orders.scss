@import "../../../sass/variables";
@import "../../../sass/mixins";

.orders__wrapper > {
    box-sizing: border-box;
}

.orders__wrapper {
    width: 590px;
    margin: 20px auto 0;
    display: flex;
    flex-direction: column;
    align-items: center;
    background-color: $color-element-background-light;
    padding: 0 0 10px 5px;
    border-radius: 5px;

    @include media-mobile {
        width: 500px;
    }

    @include media-small-mobile {
        width: 350px;
    }
}
