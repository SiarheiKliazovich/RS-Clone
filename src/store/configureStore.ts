/* eslint-disable @typescript-eslint/ban-ts-comment */
import { applyMiddleware, createStore } from "redux";
import thunkMiddleware from "redux-thunk";

import loggerMiddleware from "./logger";
import rootReducer from "../reducers";

const configureStore = () => {
  const middlewares = [loggerMiddleware, thunkMiddleware];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(rootReducer, middlewareEnhancer);

  // @ts-ignore:next-line
  if (process.env.NODE_ENV !== "production" && module.hot) {
    // @ts-ignore:next-line
    module.hot.accept("./../reducers", () => store.replaceReducer(rootReducer));
  }

  return store;
};

export default configureStore;
