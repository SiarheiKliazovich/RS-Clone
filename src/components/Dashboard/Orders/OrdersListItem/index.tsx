import { FunctionComponent, useState } from "react";
import { useSelector } from "react-redux";

import { lightTheme, nightTheme } from "../../../../data/constants";
import { IStore } from "../../../../interfaces/store";
import { IOrdersListItem } from "@/src/interfaces/ordersListItem";

const OrdersListItem: FunctionComponent<IOrdersListItem> = ({
  item,
  i,
  deleteItem,
  updateItem,
}) => {
  const appInterfaceStore = useSelector((state: IStore) => state.appInterface);
  const { isNightMode } = appInterfaceStore;

  const [selectedValue, setSelectedValue] = useState("EMPTY");

  const backgroundColor = isNightMode
    ? nightTheme.background.element
    : lightTheme.background.element;

  return (
    <>
      <tr>
        <th>{i + 1}</th>
        <th>{item.date.slice(0, 10)}</th>
        <th>{item.price}</th>
        <th>{item.status}</th>
        <th>
          <button
            style={{ backgroundColor }}
            className="icon"
            onClick={() => deleteItem(item._id)}
          ></button>
        </th>
        <th className="table-select">
          <select
            className="select"
            value={selectedValue}
            onChange={(e) => setSelectedValue(e.target.value)}
          >
            <option value="EMPTY" disabled hidden></option>
            <option value="NON-PAID">NON-PAID</option>
            <option value="PAID">PAID</option>
            <option value="DECLINED">DECLINED</option>
            <option value="IN-PROGRESS">IN PROGRESS</option>
            <option value="FINISHED">FINISHED</option>
          </select>
        </th>
        <th>
          <button
            className="button-apply"
            onClick={() => updateItem(item._id, selectedValue)}
          >
            Apply
          </button>
        </th>
      </tr>
    </>
  );
};

export default OrdersListItem;
