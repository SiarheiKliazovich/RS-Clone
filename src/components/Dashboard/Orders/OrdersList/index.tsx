import { FunctionComponent } from "react";
import { useDispatch } from "react-redux";

import {
  deleteOrder,
  getOrders,
  updateOrder,
} from "../../../../services/apiOrders";
import {
  orderDeleted,
  ordersFetching,
  ordersFetched,
  ordersFetchingError,
} from "../../../../actions";
import { IOrdersListComponent } from "../../../../interfaces/ordersListComponent";
import OrdersListItem from "../OrdersListItem";
import { useTranslation } from "react-i18next";

import "./ordersList.scss";

const OrdersList: FunctionComponent<IOrdersListComponent> = ({ orders }) => {
  const dispatch = useDispatch();

  const { t } = useTranslation("dashboard");

  const deleteItem = async (id: string) => {
    try {
      await deleteOrder(id);
      dispatch(orderDeleted(id));
      dispatch(ordersFetching());
      const orders = await getOrders();
      dispatch(ordersFetched(orders));
    } catch {
      dispatch(ordersFetchingError());
    }
  };

  const updateItem = async (id: string, status: string) => {
    try {
      await updateOrder(id, { status: status });
      dispatch(ordersFetching());
      const orders = await getOrders();
      dispatch(ordersFetched(orders));
    } catch {
      dispatch(ordersFetchingError());
    }
  };

  return (
    <>
      <div className="orders-list__wrapper">
        <div className="orders-list__item">
          <table>
            <tbody>
              <tr className="table-header">
                <th>#</th>
                <th>{t("ordersTable.date")}</th>
                <th>{t("ordersTable.price")}</th>
                <th>{t("ordersTable.status")}</th>
              </tr>
              {orders.map((item, i) => {
                return (
                  <OrdersListItem
                    key={item._id}
                    item={item}
                    i={i}
                    deleteItem={deleteItem}
                    updateItem={updateItem}
                  />
                );
              })}
            </tbody>
          </table>
        </div>
      </div>
    </>
  );
};

export default OrdersList;
