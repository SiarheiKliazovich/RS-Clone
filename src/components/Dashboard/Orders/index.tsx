import { FunctionComponent, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";

import { getOrders } from "../../../services/apiOrders";
import {
  ordersFetching,
  ordersFetched,
  ordersFetchingError,
} from "../../../actions";
import { IStore } from "../../../interfaces/store";

import { IOrdersReducer } from "../../../interfaces/ordersReducer";
import { lightTheme, nightTheme } from "../../../data/constants";
import OrdersList from "./OrdersList";
import Spinner from "../../Spinner";
import ErrorMessage from "../../ErrorMessage";

import "./orders.scss";

const Orders: FunctionComponent = () => {
  const ordersStore = useSelector(
    (state: IOrdersReducer) => state.ordersReducer
  );
  const { ordersLoadingStatus, orders } = ordersStore;
  const appInterfaceStore = useSelector((state: IStore) => state.appInterface);
  const { isNightMode } = appInterfaceStore;

  const backgroundColor = isNightMode
    ? nightTheme.background.element
    : lightTheme.background.element;

  const dispatch = useDispatch();

  const getOrdersList = async () => {
    try {
      dispatch(ordersFetching());
      const orders = await getOrders();
      dispatch(ordersFetched(orders));
    } catch {
      dispatch(ordersFetchingError());
    }
  };

  useEffect(() => {
    getOrdersList();
  }, []);

  const spinner = ordersLoadingStatus === "loading" ? <Spinner /> : null;

  return (
    <div className="orders__wrapper" style={{ backgroundColor }}>
      {spinner}
      {ordersLoadingStatus === "idle" ? <OrdersList orders={orders} /> : null}
      {ordersLoadingStatus === "error" ? <ErrorMessage /> : null}
    </div>
  );
};

export default Orders;
