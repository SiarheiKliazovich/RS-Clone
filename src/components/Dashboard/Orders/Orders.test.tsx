import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import Orders from ".";

const mockStore = configureStore([]);

describe("FavoriteTie", () => {
  // eslint-disable-next-line
  let store: any;

  beforeEach(() => {
    store = mockStore({
      appInterface: {
        accentColor: {
          static: "#000",
          hover: "#fff",
        },
      },
      ordersReducer: {
        ordersLoadingStatus: "idle",
        orders: [],
      },
    });
  });

  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <Orders />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
