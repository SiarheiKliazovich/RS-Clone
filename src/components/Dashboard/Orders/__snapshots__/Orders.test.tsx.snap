// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteTie renders correctly 1`] = `
<div>
  <div
    class="orders__wrapper"
    style="background-color: rgb(255, 255, 255);"
  >
    <div
      class="orders-list__wrapper"
    >
      <div
        class="orders-list__item"
      >
        <table>
          <tbody>
            <tr
              class="table-header"
            >
              <th>
                #
              </th>
              <th>
                ordersTable.date
              </th>
              <th>
                ordersTable.price
              </th>
              <th>
                ordersTable.status
              </th>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
